File: DataBinding/PageAndRDLX/ODataDataSource/C#/ODataEndPoint/Controllers/CustomersController.cs
using System.Collections.ObjectModel;
using Microsoft.Data.Sqlite;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.OData.Query;
using Microsoft.AspNetCore.OData.Routing.Controllers;
using ODataEndPoint.Models;

namespace ODataEndPoint.Controllers
{
	public class CustomersController : ODataController
	{
		private readonly IConfiguration _configuration;
		private readonly IWebHostEnvironment _env;

		public CustomersController(IConfiguration configuration, IWebHostEnvironment env)
		{
			_configuration = configuration;
			_env = env;
		}
		
		[EnableQuery]
		public IActionResult Get()
		{
			var connStr = _configuration.GetSection("ConnectionStrings")["Nwind"]?.Replace("$appPath$", _env.ContentRootPath);
			var conn = new SqliteConnection(connStr);
			conn.Open();
			var customers = new Collection<Customer>();
			var cmd = new SqliteCommand("select customers.Customerid, customers.CompanyName, customers.ContactName, customers.Address from customers", conn);
			var dataReader = cmd.ExecuteReader();
			while (dataReader.Read())
			{
				customers.Add(new Customer
				{
					CustomerID = dataReader.GetValue(0).ToString()!,
					CompanyName = dataReader.GetValue(1).ToString()!,
					ContactName = dataReader.GetValue(2).ToString()!,
					Address = dataReader.GetValue(3).ToString()!,
				});
			}
			conn.Close();
			return Ok(customers.AsQueryable());
		}
	}
}

File: DataBinding/PageAndRDLX/ODataDataSource/VB.NET/ODataEndPoint/Controllers/CustomersController.vb
Imports System.Collections.ObjectModel
Imports Microsoft.Data.Sqlite
Imports Microsoft.AspNetCore.Hosting
Imports Microsoft.AspNetCore.Mvc
Imports Microsoft.AspNetCore.OData.Query
Imports Microsoft.AspNetCore.OData.Routing.Controllers
Imports Microsoft.Extensions.Configuration
Imports ODataEndPoint.Models

Namespace Controllers

	Public Class CustomersController
		Inherits ODataController

		ReadOnly _configuration As IConfiguration
		ReadOnly _env As IWebHostEnvironment

		Public Sub New(configuration As IConfiguration, env As IWebHostEnvironment)
			_configuration = configuration
			_env = env
		End Sub

		<EnableQuery>
		Public Function [Get]() As IActionResult
			Dim connStr = _configuration.GetSection("ConnectionStrings")("Nwind").Replace("$appPath$", _env.ContentRootPath)
			Dim conn As New SqliteConnection(connStr)
			conn.Open()
			Dim customers As New Collection(Of Customer)()
			Dim cmd As New SqliteCommand("select customers.CustomerID, customers.CompanyName, customers.ContactName, customers.Address from customers", conn)
			Dim dataReader As SqliteDataReader = cmd.ExecuteReader()
			While dataReader.Read()
				customers.Add(New Customer() With {
								 .CustomerID = dataReader.GetValue(0).ToString(),
								 .CompanyName = dataReader.GetValue(1).ToString(),
								 .ContactName = dataReader.GetValue(2).ToString(),
								 .Address = dataReader.GetValue(3).ToString()
								 })
			End While

			conn.Close()
			Return Ok(customers.AsQueryable())
		End Function
	End Class
End NameSpace